### PR TITLE
Unify run and sample into shared tool_runner

### PR DESCRIPTION
**Pull request by marantic-amd**
## Motivation

- Both `rocprof-sys-run` and `rocprof-sys-sample` had near-identical impl.cpp files (~80% shared code) that diverged over time, causing inconsistent help output, flag names, and behaviors.

## Technical Details

- Unification

  - Extract the shared lifecycle (env init → arg parsing → env prep → exec) into a single `tool_runner`. Each binary becomes a thin `main()` that calls` run_tool(argc, argv, tool_mode)`.
  - `rocprof-sys-sample` is now what users always assumed it was - `rocprof-sys-run` configured for sampling mode - and the two tools can no longer drift accidentally.

- Bug fixes

  - `update_env(REPLACE)` no longer leaves duplicate `KEY= entries`. Previously it modified only the first match; a later `consolidate_env_entries` then joined the parts and silently turned `REPLACE into APPEND`. 
  - Concretely: 
    - `ROCPROFSYS_TRACE=true` in the shell plus a preset setting it to false produced `TRACE=false:true` (truthy), yielding a perfetto trace under `--profile-only`. 
    - `REPLACE` now writes the first match and erases subsequent matches in one pass. Same issue independently reported and patched against develop in #4620.
  - `init_parser` no longer captures the initial environment — that is the caller's job (`get_initial_environment` in `tool_runner`). The duplicate population was the source of the duplicate entries above.

- Type-safety / RAII migration

  - `parser_data::current` and `parser_data::command` moved from `std::vector<char*>` (with manual `strdup/free`, leak-prone via `original_envs` shadow set) to `std::vector<std::string>`. The `vector<char*>` shim is built only at the `execvpe` boundary via a single `utils::to_c_argv` helper.
  - `parser_data` fields regrouped into `env_snapshot` / `arse_outcome` / `registration_config` sub-structs; `env_snapshot::updated` flipped from `unordered_set<string_view>` to `unordered_set<string>` (string_view storage was fragile against temporary keys).
  - `tool_config` struct collapsed into a `tool_mode enum` once the per-mode logic settled inside `tool_runner`.

- Control flow

  - `cli_done` exception (thrown only to be caught and translated back to an exit code by `--list-presets/--explain/--help/--export-config`) replaced with `std::optional<int>` returns; explicit short-circuit in run_tool.
  - Monolithic `parse_args` split into named stages (`get_initial_environment`, `parse_command_fast_path`, `do_full_parse`, `apply_post_parse`).

- API / perf cleanups

  - `common::add_torch_library_path` now takes `std::string_view` executable instead of `std::vector<char*>` (only `argv.front()` was ever inspected).
  - `prepare_command` matches launcher against `basename(arg)` rather than substring on the full path, so launcher names don't spuriously match path components.
  - `std::regex` swapped for `std::string_view::find` in launcher matching — no behavioral change for any real launcher value, drops <regex> allocation per command-arg iteration.
  - `dispatch_help` drops dead try/catch around `parser.get<string>("help");` `replace_all` rewritten as a clearer while-loop; `print_environment` `std::unique` after sort; `check_directory_writable` walks ancestors until one exists rather than only checking the immediate parent.
  - Test fixture infrastructure (`make_env`/`m_allocated`) removed from `test_print_output` — unused since the move to `vector<string>`.

## JIRA ID

This is a needed refactor which also resolves issues mentioned in AIPROFSYST-407

## Test Plan

- Existing unit tests suite passes (covers test_update_env, test_environment, test_remove_env, test_print_output, test_help_system, test_preset_registry, test_json_config, test_discover_llvm_libdir, test_path, demangler_test).
- New regression tests in test_update_env.cpp:
  - ReplaceMode_RemovesDuplicateEntries — [KEY=A, OTHER=keep, KEY=A] → REPLACE → [KEY=B, OTHER=keep].
  - ReplaceMode_RemovesAllDuplicatesWhenManyExist — REPLACE collapses 3 duplicates to 1.
- AddTorchLibraryPathTest updated for the new string_view signature.
- Manual smoke test reproducing the original bug:
```shell
export ROCPROFSYS_TRACE=true
rocprof-sys-run    --preset=profile-only -- /tmp/hello
rocprof-sys-sample --preset=profile-only -- /tmp/hello
```
- Both tools produce only wall_clock.{txt,json} + metadata/functions JSON; no perfetto-trace.proto.

## Test Result

All existing tests pass

## Submission Checklist

- [x] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

### CODE REVIEW

File: projects/rocprofiler-systems/source/bin/common/tool_runner.hpp [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    {
+        return is_sample();
+    }
+};
+
+tool_config
+make_run_config();
+
+tool_config
+make_sample_config();
```
Comment (mradosav-amd) [inline projects/rocprofiler-systems/source/bin/common/tool_runner.hpp]: This looks like factory.. You have enum as well.
`run_tool` can take enum, and factory can create tool inside it.
It's not necessary to expose tool_config and creation of it, since it's gonna be immediately transferred inside `run_tool`.

File: projects/rocprofiler-systems/source/bin/common/common_utils.cpp [hunk 2]
```diff
@@ -22,6 +26,86 @@ namespace rocprofsys
...
+    for(auto& entry : src)
+        out.emplace_back(entry.data());
+    out.emplace_back(nullptr);
+    return out;
+}
+
+void
+print_command(const std::vector<std::string>& argv, std::string_view prefix)
+{
+    auto cmd = std::accumulate(argv.begin(), argv.end(), std::string{},
```
Comment (sputhala-amd) [inline projects/rocprofiler-systems/source/bin/common/common_utils.cpp:62]: This code is not a hot path at all so doesnt matter, but is O(n2), Can be replaced with:

```suggestion
 if(argv.empty()) return;

    size_t total = 0;
    for(const auto& arg : argv)
        total += arg.size();
    if(argv.size() > 1) total += argv.size() - 1;

    std::string cmd;
    cmd.reserve(total);
    for(const auto& arg : argv)
    {
        if(!cmd.empty()) cmd += ' ';
        cmd += arg;
    }
```

File: projects/rocprofiler-systems/source/bin/common/tool_runner.cpp [hunk 3]
```diff
@@ -0,0 +1,622 @@
...
+}
+
+std::optional<int>
+tool_runner::do_full_parse()
+{
+    toggle_suppression(pre_main_suppression_guard);
+    rocprofsys::argparse::init_parser(data);
+    signals::disable_signal_detection(signals::signal_settings::get_enabled());
+
+    auto parser = parser_t{ basename(argv[0]), build_description() };
```
Comment (sputhala-amd) [inline projects/rocprofiler-systems/source/bin/common/tool_runner.cpp]: `basename` from `<libgen.h>` takes char* and is allowed to modify the buffer. That can mutate the process’s `argv` string and break later logic that still uses `argv`. `prepare_command` has same logic in the `basename_of` lambda which we can re-use.

File: projects/rocprofiler-systems/source/bin/common/common_utils.cpp (hunk 2 above)
Comment (marantic-amd) [reply projects/rocprofiler-systems/source/bin/common/common_utils.cpp:62]: I opted for this solution because it's clearer and doesn't introduces any overhead that would matter

File: projects/rocprofiler-systems/source/bin/common/tool_runner.cpp (hunk 3 above)
Comment (marantic-amd) [reply projects/rocprofiler-systems/source/bin/common/tool_runner.cpp]: I will look deeper into it, this can mess up our flow in certain use-cases, might be really rare, but still, I wouldn't like to leave that room for mistake. Thank you

File: projects/rocprofiler-systems/source/lib/common/environment.hpp
Comment (dgaliffiAMD) [inline projects/rocprofiler-systems/source/lib/common/environment.hpp]: This was pre-existing, but I don't see a check for `parts.empty() `. As such, this may underflow if `parts` is empty.

File: projects/rocprofiler-systems/source/bin/common/tool_runner.hpp [hunk 4]
```diff
@@ -0,0 +1,20 @@
+// Copyright (c) Advanced Micro Devices, Inc.
```
Comment (mradosav-amd) [inline projects/rocprofiler-systems/source/bin/common/tool_runner.hpp:1]: can we write unit tests for this module?

Comment (marantic-amd) [reply projects/rocprofiler-systems/source/bin/common/tool_runner.hpp:1]: We can, but it would mean exposing additionally some of the helper functionalities just for the sake of testing. What this module does is almost pure orchestration, and all of the bottleneck parts are already covered in the test of the corresponding module. 

I noticed that I have missed maybe some edge case E2E tests, so I will add those, but regarding unit tests, I think for this module it wouldn't make much sense

**Comment by sputhala-amd** [comment]
LGTM. Need to resolve merge conflicts.

**Comment by dgaliffiAMD** [comment]
Looks good to me. 
Please update the `CHANGELOGS` noting that, for `rocprof-sys-sample` some arguments have been renamed to align with the other tools. 
- `--freq` to `--sampling-freq` 
- `cputime` and `realtime` to `--sample-cputime` and `--sample-realtime`, respectively.

**Review by dgaliffiAMD — APPROVED** [review]
